core/model/modx/processors/security/login.class.php

```diff
@@ -140,23 +140,29 @@ public function checkIsBlocked() {
         /** @var modUserProfile $profile */
         $profile = $this->user->Profile;
 
-        if ($profile->get('failed_logins') >= $this->modx->getOption('failed_login_attempts') &&
-            $profile->get('blockeduntil') > time()) {
-            return $this->modx->lexicon('login_blocked_too_many_attempts');
-        }
-
-        if ($profile->get('failedlogincount') >= $this->modx->getOption('failed_login_attempts')) {
-            $profile->set('failedlogincount', 0);
-            $profile->set('blocked', 1);
-            $profile->set('blockeduntil', time() + (60 * $this->modx->getOption('blocked_minutes')));
+        /* Update block state */
+        if ($profile->get('blockeduntil') < time()) {
+            if ($profile->get('blockeduntil') > 0) {
+                $profile->set('failedlogincount', 0);
+            }
+            $profile->set('blockeduntil', 0);
             $profile->save();
         }
-        if ($profile->get('blockeduntil') != 0 && $profile->get('blockeduntil') < time()) {
+        if ($this->user->passwordMatches($this->givenPassword)) {
```

Comment:
I just ran some tests. If I try to login a few times with a wrong password I get the `login_blocked_too_many_attempts` message as expected.

If I login a few more times with a wrong password the message stays the same.
But when I login after `login_blocked_too_many_attempts` with the correct password I can't login because I'm blocked (as expected). 

When I try to login again with a wrong/correct password the message I receive is different (`login_blocked_error`). This happens because after the password matches `failedlogincount` is reset to 0.

An attacker can use this behaviour to retrieve a users password.

Reply:
@minagerges do you have time to check my comment?